cni/pkg/install-cni/pkg/install/install.go

```diff
@@ -0,0 +1,265 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package install
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+
+	"istio.io/istio/cni/pkg/install-cni/pkg/config"
+	"istio.io/istio/cni/pkg/install-cni/pkg/constants"
+	"istio.io/istio/cni/pkg/install-cni/pkg/util"
+	"istio.io/pkg/log"
+)
+
+type configFiles struct {
+	kubeconfigFilepath string
+	cniConfigFilepath  string
+}
+
+// Run starts the installation process with given configuration
+func Run(cfg *config.Config) (err error) {
+	var files configFiles
+
+	ctx, cancel := context.WithCancel(context.Background())
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+	go func(sigChan chan os.Signal, cancel context.CancelFunc) {
+		sig := <-sigChan
+		log.Infof("Exit signal received: %s", sig)
+		cancel()
+	}(sigChan, cancel)
+
+	defer func() {
+		if cleanErr := cleanup(cfg, files); cleanErr != nil {
+			if err != nil {
+				err = errors.Wrap(err, cleanErr.Error())
+			} else {
+				err = cleanErr
+			}
+		}
+	}()
+
+	for {
+		if err = copyBinaries(cfg.UpdateCNIBinaries, cfg.SkipCNIBinaries); err != nil {
+			return
+		}
+
+		var saToken string
+		saToken, err = readServiceAccountToken()
+		if err != nil {
+			return
+		}
+
+		files.kubeconfigFilepath, err = createKubeconfigFile(cfg, saToken)
+		if err != nil {
+			return
+		}
+
+		files.cniConfigFilepath, err = createCNIConfigFile(ctx, cfg, saToken)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				// Error was caused by interrupt/termination signal
+				err = nil
+			}
+			return
+		}
+
+		if !cfg.Sleep {
+			err = checkInstall(cfg, files.cniConfigFilepath)
+			return
+		}
+		// Otherwise, keep container alive
+
+		// Create file watcher before checking for installation
+		// so that no file modifications are missed while and after checking
+		var watcher *fsnotify.Watcher
+		var fileModified chan bool
+		var errChan chan error
+		watcher, fileModified, errChan, err = util.CreateFileWatcher(cfg.MountedCNINetDir)
+		if err != nil {
+			return
+		}
+
+		for {
+			if checkErr := checkInstall(cfg, files.cniConfigFilepath); checkErr != nil {
+				log.Infof("Invalid configuration. %v", checkErr)
+				break
+			}
+			// Valid configuration; Wait for modifications before checking again
+			if waitErr := util.WaitForFileMod(ctx, fileModified, errChan); waitErr != nil {
+				if !errors.Is(waitErr, context.Canceled) {
+					// Error was not caused by interrupt/termination signal
+					err = waitErr
+				}
+				if closeErr := watcher.Close(); closeErr != nil {
+					if err != nil {
+						err = errors.Wrap(err, closeErr.Error())
+					} else {
+						err = closeErr
+					}
+				}
+				return
+			}
+		}
+
+		if err = watcher.Close(); err != nil {
```

Comment:
Is it better just log the error and continue the loop instead of aborting?

Reply:
Good point, it doesn't seem like it should be a terminating error. I'll change it to log.
Also it seems that the [`watcher.Close()` function](https://github.com/fsnotify/fsnotify/blob/7f4cf4dd2b522a984eaca51d1ccee54101d3414a/inotify.go#L73) doesn't ever return a non-nil error?